src/main/java/io/javalin/core/JavalinServlet.kt

```diff
@@ -107,26 +106,36 @@ class JavalinServlet(
                     }
                     tryErrorHandlers()
                     tryAfterHandlers()
-                    writeResult(ctx, asyncContext.response as HttpServletResponse)
+                    writeResult(ctx, asyncContext.response as HttpServletResponse, type)
+                    logRequest(ctx)
                     asyncContext.complete()
                 }
             }
         }
     }
 
-    private fun writeResult(ctx: Context, res: HttpServletResponse) {
-        if (!res.isCommitted) {
-            ctx.resultStream()?.let { resultStream ->
-                if (gzipShouldBeDone(ctx)) {
-                    GZIPOutputStream(res.outputStream, true).use { gzippedStream ->
-                        res.setHeader(Header.CONTENT_ENCODING, "gzip")
-                        resultStream.copyTo(gzippedStream)
-                    }
-                } else {
-                    resultStream.copyTo(res.outputStream)
-                }
+    private fun writeResult(ctx: Context, res: HttpServletResponse, type: HandlerType) {
+        if (res.isCommitted || ctx.resultStream() == null) return // nothing to write
+        val resultStream = ctx.resultStream()!!
+        if (res.getHeader(Header.ETAG) != null || (autogeneratedEtagsEnabled && type == HandlerType.GET)) {
```

Comment:
you don't make ETags for other than GETs? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match mentions HEAD as well. I have no idea.

Reply:
Not auto-generated. Users can still set etag for any request type.